### PR TITLE
[very early WIP] add openssl/3.0.0

### PR DESCRIPTION
**Pull request by Croydon**
Specify library name and version:  **openssl/3.0.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

_____________________________

The intention of this is **NOT**:
  * to merge any pre-release versions - all pre-release versions will get removed later
  * to merge any of this before the final 3.0.0 version gets released, which means this will stay a draft for several more months

The intention of this is:
  * to get early feedback
  * to test early on and notice bugs early on
  * to be able to release a recipe for a final 3.0.0 version shortly after its release
  * to get contributions (help wanted!) 😄 

Important resources:
  * https://wiki.openssl.org/index.php/OpenSSL_3.0
  * https://github.com/openssl/openssl/blob/master/NEWS.md#openssl-30
  * https://github.com/openssl/openssl/blob/master/INSTALL.md

Upstream bugs:
  * [x] Builds fails when path contains `@` https://github.com/openssl/openssl/issues/12078

Open questions / TODO (certainly incomplete):
  * [ ] since #1586 we patched OpenSSL for some Apple operating systems which don't have `fork()`. OpenSSL merged in the meantime a "partial fix" (https://github.com/openssl/openssl/commit/ca811248d838058c13236a6c3b688e0ac98c02c8), so even when we set one of the flags (`HAVE_FORK` or `NO_FORK`, which one would it be anyway? Why two of them?) it does not seem to be fixed upstream. But we should probably wait a few pre-releases longer, before we might patch from our side again. See also https://github.com/openssl/openssl/issues/7607
  * [ ] in `package()` when `nmake` is used `build_type` is `Debug` we give the lib files a `d` prefix. Why? Is this really wanted and still needed for >= 3.0.0?
  * [ ] we seem to have several problems with `find_package(OpenSSL)`/`FindOpenSSL.cmake`, see my comments here: https://github.com/conan-io/conan-center-index/pull/2054/commits/a2bb89ba04d90df0d01c3e34daf21485a5cae5d1
  * [ ] the `test_package` will probably fail in its current form with certain different option values
  * [ ] Usage of environment variable `CROSS_COMPILE` https://github.com/conan-io/conan-center-index/pull/2054#discussion_r523453297

### CODE REVIEW

File: recipes/openssl/3.x.x/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+        return self._is_clangcl or self._is_msvc
+
+    def source(self):
+        try:
+            tools.get(**self.conan_data["sources"][self.version])
+        except ConanException:
+            self.output.warn("Downloading OpenSSL from the mirror.")
+            url = self.conan_data["sources"][self.version]["url"]
+            # FIXME: For < 3.0.0 the old url contained major.minor.patch, e.g. 1.1.0, but without the letter (e.g. "a")
+            # FIXME: So what will be the old url for >= 3.0.0?
```
Comment (prince-chrismc) [inline recipes/openssl/3.x.x/conanfile.py:88]: What if we used github for the sources?

Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:88]: They don't upload their official release sources on GitHub. We only have there the auto generated ones, which is really bad when it comes to something security sensitive like OpenSSL.

OpenSSL provides official hash sums which then would not match anymore

File: recipes/openssl/3.x.x/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+        if not make_program:
+            raise Exception('could not find "make" executable. please set "CONAN_MAKE_PROGRAM" environment variable')
+        return make_program
+
+    def _patch_install_name(self):
+        if self.settings.os == "Macos" and self.options.shared:
+            old_str = '-install_name $(INSTALLTOP)/$(LIBDIR)/'
+            new_str = '-install_name '
+
+            tools.replace_in_file("Makefile", old_str, new_str, strict=self.in_local_cache)
```
Comment (prince-chrismc) [inline recipes/openssl/3.x.x/conanfile.py:556]: This failed in the build 12, seems like it was moved to a perl script. https://github.com/openssl/openssl/blame/7f81aed4bccbeb0ad491b68dd601d576c4956c78/Configurations/shared-info.pl#L46

Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:556]: This is weird. The commit change this part is already part of 1.1.1

https://github.com/openssl/openssl/commit/793077d0beccfa20c9962546393128b92a7e68e4

**Comment by madebr** [comment]
We merged #3558 today that adds a lot of new options.
Would it be possible to add these option to 3.x as well?

File: recipes/openssl/3.x.x/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+            if not self._win_bash:
+                self.build_requires("strawberryperl/5.30.0.1")
+            if not self.options.no_asm and not tools.which("nasm"):
+                self.build_requires("nasm/2.14")
+        if self._win_bash:
+            if "CONAN_BASH_PATH" not in os.environ:
+                self.build_requires("msys2/20200517")
+
+    def requirements(self):
+        if self.options.get_safe("zlib") == True:
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:66]:
```suggestion
        if self.options.zlib:
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+            if self.options.enable_capieng:
+                args.append("enable-capieng")
+            if self.options.capieng_dialog:
+                args.append("-DOPENSSL_CAPIENG_DIALOG=1")
+        else:
+            args.append("-fPIC" if self.options.fPIC else "no-pic")
+        if self.settings.os == "Neutrino":
+            args.append("-lsocket no-asm")
+
+        if self.options.get_safe("zlib") == True:
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:369]:
```suggestion
        if self.options.zlib:
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+            libsuffix = "d" if self.settings.build_type == "Debug" else ""
+            self.cpp_info.components["ssl"].libs = ["libssl" + libsuffix]
+            self.cpp_info.components["crypto"].libs = ["libcrypto" + libsuffix]
+        else:
+            self.cpp_info.components["ssl"].libs = ["ssl"]
+            self.cpp_info.components["crypto"].libs = ["crypto"]
+
+        self.cpp_info.components["ssl"].requires = ["crypto"]
+
+        if self.options.get_safe("zlib"):
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:613]:
```suggestion
        if self.options.zlib:
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+    def _patch_makefile_org(self):
+        # https://wiki.openssl.org/index.php/Compilation_and_Installation#Modifying_Build_Settings
+        # its often easier to modify Configure and Makefile.org rather than trying to add targets to the configure scripts
+        def adjust_path(path):
+            return path.replace("\\", "/") if tools.os_info.is_windows else path
+
+        makefile_org = os.path.join(self._source_subfolder, "Makefile.org")
+        env_build = self._get_env_build()
+        with tools.environment_append(env_build.vars):
+            if not "CROSS_COMPILE" in os.environ:
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:308]: The use of `CROSS_COMPILE` looks wrong to me.
I think it should test for `tools.cross_building(self.settings, skip_x86_64=True)`.

File: recipes/openssl/3.x.x/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+            self._run_make()
+
+    def _make_install(self):
+        with tools.chdir(self._source_subfolder):
+            self._run_make(targets=["install_sw"], parallel=False)
+
+    @property
+    def _cc(self):
+        if "CROSS_COMPILE" in os.environ:
+            return "gcc"
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:526]: This looks wrong too.
I would remove this.

File: recipes/openssl/3.x.x/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+        "no_md5": [True, False],
+        "no_mdc2": [True, False],
+        "no_rc2": [True, False],
+        "no_rc4": [True, False],
+        "no_rsa": [True, False],
+        "no_sha": [True, False],
+        "no_sse2": [True, False],
+        "no_threads": [True, False],
+        "openssldir": "ANY",
+        "zlib": [True, False]
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:45]: This option was named `no_zlib` in 1.x.
If you intended to rename it, maybe go for `with_zlib`?

Comment (madebr) [reply recipes/openssl/3.x.x/conanfile.py:45]: Also, in 1.1 this option was removed.
Has it come back? Was it ever removed?

File: recipes/openssl/3.x.x/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def config_options(self):
+        if self.settings.os != "Windows":
+            del self.options.capieng_dialog
+            del self.options.enable_capieng
+        else:
+            del self.options.fPIC
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:94]: swap `config_options` and `configure` + remove `fPIC` when shared.

```suggestion
    def config_options(self):
        if self.settings.os != "Windows":
            del self.options.capieng_dialog
            del self.options.enable_capieng
        else:
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            del self.options.fPIC
        del self.settings.compiler.libcxx
        del self.settings.compiler.cppstd
```

File: recipes/openssl/3.x.x/conanfile.py (hunk 8 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:45]: It looks to me like the upstream option `no_zlib` was last available in 1.0.1 https://github.com/openssl/openssl/blob/OpenSSL_1_0_1g/INSTALL#L60

then got removed in >= 1.1.1

However, a `zlib` option was always available

https://github.com/openssl/openssl/blob/OpenSSL_1_1_1h/INSTALL#L550
https://github.com/openssl/openssl/blob/master/INSTALL.md#zlib

That is where the name is coming from. I followed upstream

File: recipes/openssl/3.x.x/conanfile.py (hunk 6 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:310]: I think this is a very ancient leftover which is worth to update in 1.x.x too

So let's spin this out in an own issue and then port the changes later

**Comment by jgsogo** [comment]
It looks like this PR has conflicts with `master` branch. They don't show up because it is a draft... and Jenkins won't run if the PR _is not mergeable_. I'll move it from draft to ready and back to see if the conflicts appear and the CI runs.

**Comment by Croydon** [comment]
I don't always update this PR instantly, but rather wait a bit and port several PRs which got merged to the 1.x version in the mean time + address PR reviews

**Comment by Croydon** [comment]
Woop. Wait. Can I get this reopened, please? @uilianries @SSE4 (or anyone else)

Not sure why I didn't mark it as a draft

**Comment by madebr** [comment]
@Croydon 
openssl 3.0.0 has been released today.
Time to update this pr and bring it back alive..

File: recipes/openssl/3.x.x/conandata.yml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+sources:
+  3.0.0-alpha11:
+    sha256: 2a18f18df6a7ba33cfcc423b77d93990bf70939c06aa2b599b1eabf6e222ea74
+    url: https://www.openssl.org/source/openssl-3.0.0-alpha11.tar.gz
```
Comment (madebr) [inline recipes/openssl/3.x.x/conandata.yml:4]:
```suggestion
  3.0.0:
    sha256: 59eedfcb46c25214c9bd37ed6078297b4df01d012267fe9e9eee31f61bc70536
    url: [
      "https://www.openssl.org/source/openssl-3.0.0.tar.gz",
      "https://www.openssl.org/source/old/openssl-3.0.0.tar.gz"
    ]
```

File: recipes/openssl/config.yml [hunk 11]
```diff
@@ -23,3 +23,5 @@ versions:
     folder: "1.x.x"
   1.1.1i:
     folder: "1.x.x"
+  3.0.0-alpha11:
```
Comment (madebr) [inline recipes/openssl/config.yml:26]:
```suggestion
  3.0.0:
```

File: recipes/openssl/3.x.x/test_package/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        cmake.definitions["OPENSSL_USE_STATIC_LIBS"] = not self.options["openssl"].shared
+        if self.settings.compiler == 'Visual Studio':
+            cmake.definitions["OPENSSL_MSVC_STATIC_RT"] = 'MT' in str(
+                self.settings.compiler.runtime)
+
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
```
Comment (madebr) [inline recipes/openssl/3.x.x/test_package/conanfile.py:26]:
```suggestion
        if not tools.cross_building(self):
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,632 @@
...
+from conans.tools import Version
+
+
+class OpenSSLConan(ConanFile):
+    name = "openssl"
+    settings = "os", "compiler", "arch", "build_type"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/openssl/openssl"
+    license = "Apache-2.0"
+    topics = ("conan", "openssl", "ssl", "tls", "encryption", "security")
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:15]:
```suggestion
    topics = ("openssl", "ssl", "tls", "encryption", "security")
```

**Comment by Croydon** [comment]
I can't open this PR myself 🙈

**Comment by Croydon** [comment]
I can't open this PR myself 🙈 Help @SSE4 @uilianries 😄

File: recipes/openssl/3.x.x/conanfile.py (hunk 13 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:15]: Done 69c1de15ee86dbd97f09213782b0bdbfdd9fdf44

File: recipes/openssl/3.x.x/test_package/conanfile.py (hunk 12 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/test_package/conanfile.py:26]: Done 69c1de15ee86dbd97f09213782b0bdbfdd9fdf44

File: recipes/openssl/3.x.x/conandata.yml (hunk 10 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conandata.yml:4]: No need for the `old` mirror anymore, btw. They have now permanent download URLs (except for pre-releases)

File: recipes/openssl/3.x.x/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,632 @@
...
+
+    def build_requirements(self):
+        if tools.os_info.is_windows:
+            if not self._win_bash:
+                self.build_requires("strawberryperl/5.30.0.1")
+            if not self.options.no_asm and not tools.which("nasm"):
+                self.build_requires("nasm/2.15.05")
+        if self._win_bash:
+            if "CONAN_BASH_PATH" not in os.environ:
+                self.build_requires("msys2/20200517")
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:63]:
```suggestion

    exports_sources = ["patches/*"]

    _env_build = None
    
    @property
    def _source_subfolder(self):
        return  "source_subfolder"

    @property
    def _settings_build(self):
        return getattr(self, "settings_build", self.settings)

    def build_requirements(self):
        if self._settings_build.os == "Windows":
            if not self._win_bash:
                self.build_requires("strawberryperl/5.30.0.1")
            if not self.options.no_asm and not tools.which("nasm"):
                self.build_requires("nasm/2.15.05")
        if self._win_bash:
            if not tools.get_env("CONAN_BASH_PATH"):
                self.build_requires("msys2/cci.latest")
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,632 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_folder = "openssl-" + self.version
+        os.rename(extracted_folder, self._source_subfolder)
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:101]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version]
                  destination=self._source_subfolder, strip_root=True)
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,632 @@
...
+
+        for option_name in self.options.values.fields:
+            activated = getattr(self.options, option_name)
+            if activated and option_name not in ["fPIC", "openssldir", "capieng_dialog", "enable_capieng"]:
+                self.output.info("activated option: %s" % option_name)
+                args.append(option_name.replace("_", "-"))
+        return args
+
+    def _create_targets(self):
+        config_template = """{targets} = (
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:402]: For a nicer layout, let's wrap this in `textwrap.dedent`.

**Comment by SSE4** [comment]
I can't re-open either. it says "The openssl-3.0 branch was force-pushed or recreated".

**Comment by Croydon** [comment]
-----> https://github.com/conan-io/conan-center-index/pull/7229

File: recipes/openssl/3.x.x/conanfile.py (hunk 15 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:101]: Done

File: recipes/openssl/3.x.x/conanfile.py (hunk 14 above)
Comment (Croydon) [reply recipes/openssl/3.x.x/conanfile.py:63]: Done a0ebb18370eba7dd7b91fa141ebabcdf020449c0

**Pull request closed without merging**